File: InteropShowcase/Native/InteropShowcaseLib/callbacks.cpp
#include "pch.h"
#include <cstdio>
#include "callbacks.h"


// This source file implements the native side of the Reverse P/Invoke bridge.
// Managed code (via NativeAOT) provides an unmanaged function pointer that
// native code can store and invoke at any time. This enables event-style
// communication from C++ back into C#.
 

/// <summary>
/// Holds the managed callback function pointer registered from C#.
/// Once set, native code can invoke this callback at any time.
/// </summary>
static managed_callback_t g_callback = nullptr;


/// <summary>
/// Calls the registered managed callback with a fixed message.
/// Demonstrates a simple native → managed invocation.
/// </summary>
EXP32 void native_do_work()
{
  if (g_callback)
    g_callback("Hello World from Native C++!");
}


/// <summary>
/// Calls the registered managed callback with a caller‑provided message.
/// This allows native code to forward arbitrary UTF‑8 text to managed code.
/// </summary>
/// <param name="msg">UTF‑8 encoded message string.</param>
EXP32 void native_do_work_msg(const char* msg)
{
  if (g_callback)
    g_callback(msg);
}


/// <summary>
/// Registers a managed callback function pointer.  
/// The pointer must reference a method compiled with UnmanagedCallersOnly,
/// ensuring it is a valid native entry point callable from C++.
/// </summary>
/// <param name="cb">Managed callback function pointer.</param>
EXP32 void register_managed_callback(managed_callback_t cb)
{
  g_callback = cb;
}

File: InteropShowcase/Managed/TestInteropShowcase/Native/CallBacks/Callbacks.cs

  
using System.Runtime.InteropServices;

namespace michele.natale.RingBufferNative;
 

/// <summary>
/// Provides managed wrappers for native callback functions.
/// This class encapsulates all Reverse P/Invoke entry points used to:
/// export unmanaged function pointers from C# (NativeAOT),
/// register them inside native code, and trigger native operations
/// that call back into managed code.
/// </summary>
internal static partial class Callbacks
{
  /// <summary>
  /// Name of the native library that exposes the callback functions.
  /// </summary>
  private const string DllName = "InteropShowcaseLib.dll";

  /// <summary>
  /// Calls the native function <c>native_do_work</c>.
  /// This function triggers a previously registered managed callback.
  /// </summary>
  [LibraryImport(DllName, EntryPoint = "native_do_work")]
  internal static partial void NativeDoWork();

  /// <summary>
  /// Calls the native function <c>native_do_work_msg</c> and passes
  /// a UTF‑8 encoded string to native code.  
  /// The native function then invokes the registered managed callback
  /// with the provided message.
  /// </summary>
  /// <param name="msg">The UTF‑8 text to send to the native function.</param>
  [LibraryImport(DllName, StringMarshalling = StringMarshalling.Utf8, EntryPoint = "native_do_work_msg")]
  internal static partial void NativeDoWork(string msg);

  /// <summary>
  /// Registers a managed callback inside native code.  
  /// The provided function pointer must be a true unmanaged function pointer,
  /// typically created via <c>delegate* unmanaged&lt;...&gt;</c> referencing a method
  /// marked with <see cref="UnmanagedCallersOnlyAttribute"/>.
  /// </summary>
  /// <param name="fn">A function pointer to an unmanaged C# method.</param>
  [LibraryImport(DllName, EntryPoint = "register_managed_callback")]
  internal static partial void RegisterCallback(IntPtr fn);
}
